cmd/clusterctl/cmd/completion.go

```diff
@@ -0,0 +1,240 @@
+# limitations under the License.
+`
+
+var (
+	completionLong = LongDesc(`
+		Output shell completion code for the specified shell (bash or zsh).
+		The shell code must be evaluated to provide interactive completion of
+		clusterctl commands. This can be done by sourcing it from the
+		.bash_profile.
+
+		Note: this requires the bash-completion framework.
+
+		To install it on Mac use homebrew:
+		    $ brew install bash-completion
+		Once installed, bash_completion must be evaluated. This can be done by
+		adding the following line to the .bash_profile
+		    $ source $(brew --prefix)/etc/bash_completion
+
+		If bash-completion is not installed on Linux, please install the
+		'bash-completion' package via your distribution's package manager.
+
+		Note for zsh users: [1] zsh completions are only supported in versions of zsh >= 5.2`)
+
+	completionExample = Examples(`
+		# Install bash completion on a Mac using homebrew
+		brew install bash-completion
+		printf "\n# Bash completion support\nsource $(brew --prefix)/etc/bash_completion\n" >> $HOME/.bash_profile
+		source $HOME/.bash_profile
+
+		# Load the clusterctl completion code for bash into the current shell
+		source <(clusterctl completion bash)
+
+		# Write bash completion code to a file and source it from .bash_profile
+		clusterctl completion bash > ~/.kube/clusterctl_completion.bash.inc
+		printf "\n# clusterctl shell completion\nsource '$HOME/.kube/clusterctl_completion.bash.inc'\n" >> $HOME/.bash_profile
+		source $HOME/.bash_profile
+
+		# Load the clusterctl completion code for zsh[1] into the current shell
+		source <(clusterctl completion zsh)`)
+
+	completionCmd = &cobra.Command{
+		Use:       "completion SHELL",
+		Short:     "Output shell completion code for the specified shell (bash or zsh)",
+		Long:      LongDesc(completionLong),
+		Example:   completionExample,
+		Args:      cobra.ExactArgs(1),
+		RunE:      runCompletion,
+		ValidArgs: GetSupportedShells(),
+	}
+)
+
+var (
+	completionShells = map[string]func(cmd *cobra.Command) error{
+		"bash": runCompletionBash,
+		"zsh":  runCompletionZsh,
+	}
+)
+
+// GetSupportedShells returns a list of supported shells
+func GetSupportedShells() []string {
+	shells := []string{}
+	for s := range completionShells {
+		shells = append(shells, s)
+	}
+	return shells
+}
+
+func init() {
+	RootCmd.AddCommand(completionCmd)
+}
+
+func runCompletion(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return errors.New("Shell not specified.")
+	}
+
+	if len(args) > 1 {
+		return errors.New("Too many arguments. Expected only the shell type.")
+	}
+
+	run, found := completionShells[args[0]]
+	if !found {
+		return fmt.Errorf("Unsupported shell type %q.", args[0])
+	}
+
+	return run(cmd.Parent())
+}
+
+func runCompletionBash(cmd *cobra.Command) error {
+	cmd.GenBashCompletion(os.Stdout)
+	return nil
+}
+
+func runCompletionZsh(cmd *cobra.Command) error {
+	zshHead := "#compdef clusterctl\n"
```

Comment:
Nit but I'd make these all static strings into constants outside the method body

Reply:
Fixed by https://github.com/kubernetes-sigs/cluster-api/pull/3622/commits/45842cee955acfbfdd7998b1e2240bdca09d3a00.